field/TinyLang/Field/Uni.hs

```diff
@@ -73,20 +73,52 @@ type SomeUni f      = Some (Uni f)
 
 deriving instance Show (Uni f a)
 deriving instance Eq   (Uni f a)
+-- NOTE:  This instance fixes the uni (similarly to Eq)
+deriving instance Ord  (Uni f a)
 
-deriving instance Show (UniVar f a)
 deriving instance Show (SomeUniVar f)
 
--- This doesn't type check:
---
--- > UniConst _ x1 == UniConst _ x2 = x1 == x2
---
--- because it requires the type of @x1@ and @x2@ to have an @Eq@ instance.
--- We could provide a similar to 'withGeqUni' combinator that can handle this situation,
--- but then it's easier to just pattern match on universes.
-instance Eq f => Eq (UniVar f a) where
-    UniVar _ v1 == UniVar _ v2 = v1 == v2
-
+instance Eq (SomeUni f) where
+    Some u1 == Some u2 = withGeqUni u1 u2 False True
+
+instance Ord (SomeUni f) where
+    Some Bool   `compare` Some Bool   = EQ
+    Some Bool   `compare` Some Field  = LT
+    Some Bool   `compare` Some Vector = LT
+    Some Field  `compare` Some Bool   = GT
+    Some Field  `compare` Some Field  = EQ
+    Some Field  `compare` Some Vector = LT
+    Some Vector `compare` Some Bool   = GT
+    Some Vector `compare` Some Field  = GT
+    Some Vector `compare` Some Vector = EQ
+
+-- NOTE:  We require that the type of variable matches
+instance Eq (SomeUniVar f) where
+    Some (UniVar u1 v1) == Some (UniVar u2 v2) = withGeqUni u1 u2 False (v1 == v2)
+
+-- NOTE:  We compare unis, if they are the same, we compare variables
+instance Ord (SomeUniVar f) where
+    Some (UniVar u1 v1) `compare` Some (UniVar u2 v2) =
+        withGeqUni u1 u2 (Some u1 `compare` Some u2) (v1 `compare` v2)
+
+-- NOTE:  Technically, we could use an automatically derived instance and push
+-- the @Eq a@ to `Typed.Core.Expr`
+-- NOTE: UniConst _ x1 == UniConst x2 will not work, as we need an Eq a
+-- instance.
+instance Eq f => Eq (UniConst f a) where
+    UniConst Bool   b1 == UniConst Bool   b2 = b1 == b2
+    UniConst Field  f1 == UniConst Field  f2 = f1 == f2
+    UniConst Vector v1 == UniConst Vector v2 = v1 == v2
+
+-- NOTE: We explicitly match on all universes in False case to ensure that the
+-- compiler will report an error when a new Uni is added.
+instance Eq f => Eq (SomeUniConst f) where
+    Some (UniConst Bool b1)   == Some (UniConst Bool b2)   = b1 == b2
+    Some (UniConst Field f1)  == Some (UniConst Field f2)  = f1 == f2
+    Some (UniConst Vector v1) == Some (UniConst Vector v2) = v1 == v2
+    Some (UniConst Bool _)    == _                         = False
+    Some (UniConst Field _)   == _                         = False
+    Some (UniConst Vector _)  == _                         = False
```

Comment:
We should extract [this](https://github.com/input-output-hk/plutus/blob/master/language-plutus-core/src/Language/PlutusCore/Universe/Core.hs) into a library one day. There this instance is defined once and for all as

```haskell
instance GEq f => Eq (Some f) where
    Some a1 == Some a2 = a1 `defaultEq` a2
```

(where [`GEq`](http://hackage.haskell.org/package/dependent-sum-0.5/docs/Data-GADT-Compare.html#t:GEq) is from here and is derivable)